Spec updates 3 (#9)
alchemyplatform/EIPs: EIPS/eip-6900.md

> Modular Smart Contract Accounts MUST implement the `IAccount` interface from [ERC-4337](./eip-4337.md).

Reviewer: unsure if it has been brought up yet but our impl is currently not conforming to this right now (IAccount requires native implementation for validateUserOp)

Author: We inherit BaseAccount which does this no?

Reviewer: ah youre right we're doing _validateSignature. i forgot that the owner plugins use validateUserOpSignature instead of validateUserOp

}
```

> Variables of the type `FunctionReference` MUST be interpreted as an address in the first 20 bytes and a function selector or function ID (as used in `IPlugin`) in the remaining 4 bytes.

Author: See comment below about why this definition was expanded. The original meaning of `FunctionReference` really only applies to execution functions now. Maybe it'd be better not to overload the meaning of `FunctionReference` here...

> Variables of the type `HookAssignment` MAY be used in the following manner to store and retrieve different types of hooks:

Author: Ah this was because (and I also wanted to discuss):

1. The only place `HookAssignment` is exposed in the spec is via `IPluginLoupe` which is also optional (I know this doesn't mean the above has to be MAY).
2. I sort of like the idea of not exposing in our interfaces how we store and find the different types of hooks in our account, and leaving that up to each account implementation instead.

I wanted to propose potentially updating IPluginLoupe to expose different methods for each of these cases without having to have the caller construct these specific HookAssignment arguments.

What do you think?

> Variables of the type `HookAssignment` MAY be used in the following manner to store and retrieve different types of hooks:

Reviewer 2: Ah yes, I see. Since all of these relationships are configured by the declaration fields of the manifest itself, we don't need to surface `HookAssignent` at all. Ok, makes sense. We then just need to rework the loupe functions to not reference this anymore.

> Variables of the type `HookAssignment` MAY be used in the following manner to store and retrieve different types of hooks:

Author: Ok cool, I'll make a note to remove this section and update IPluginLoupe. I also realized in the process that I forgot to expose a way to set plugin specific hooks for `executeFromPlugin` flows in the manifest. Rethinking quickly whether we really need this...

- Packed function selector and a validator `FunctionReference`, with the last byte set to 0.
  - For pre user operation validation hooks, pre runtime validation hooks, and execution hooks on standard execution functions.
- Packed function selector and a plugin address, with the last byte set to 1.
  - For execution hooks that apply only to the `executeFromPlugin` and `executeFromPluginExternal` flows.
- Packed function selector, with the last byte set to 2.
  - For execution hooks that apply to all calls to the account that match the selector except those of standard execution functions.

#### `IPluginManager.sol`

> Plugin manager interface. Modular Smart Contract Accounts SHOULD implement this interface to support installing and uninstalling plugins.

Author: Just wanted to provide a bit of flexibility in how accounts choose to define their interface for installing plugins.

--

For your second point:

I separated the plugin execution functions from `IStandardExecutor` for a couple reasons:
1. Clarity - we refer to standard execution functions and include them within an eponymous interface. If we also include the plugin execution functions in the same interface, it could lead to confusion.
2. We have language below under "Calls to execution functions" that states: 
> Alternatively, when the modular account natively implements functions in `IPluginManager` and `IStandardExecutor`, the same following steps MUST be performed for those functions. Other native functions MAY perform these steps.

This rule doesn't apply to plugin execution functions as they should not have validators and separate hooks.

Side note: Maybe the `executeFromPluginExternal` function should allow hooks?

> Plugin manager interface. Modular Smart Contract Accounts SHOULD implement this interface to support installing and uninstalling plugins.

Author: Were you proposing that we go with `MUST`? I think I'm actually okay with that, given that the IPlugin interface is a `MUST` and how we install is tightly coupled with how the plugin manifest handles dependencies.

> Plugin manager interface. Modular Smart Contract Accounts SHOULD implement this interface to support installing and uninstalling plugins.

Reviewer 2: Yeah leaning towards `MUST`. Initially, we kept this optional in case users wanted "immutable" accounts in a similar config style to immutable diamond proxies, but that's seeming like a sufficiently small use case for implementors to be OK with spec-divergence.

> Plugin manager interface. Modular Smart Contract Accounts SHOULD implement this interface to support installing and uninstalling plugins.

Reviewer: wanted to note that "SHOULD" allows a safe{core} type architecture since those functions live in the manager. that said, i dont know if we care about that at all

```solidity
interface IPluginManager {
    struct PluginInstallConfig {
        address plugin;
        bytes32 manifestHash;
        bytes data;
        address[] dependencies;
    }

    struct PluginUninstallConfig {
        address plugin;
        bytes data;

>     }

Author: It was also a way to be able to potentially specify init data as an array of PluginInstallConfig, instead of having to pass an array of each member and do array length checking. But seeing as how we're doing that in the factory now anyway, perhaps this makes sense.

    event PluginInstalled(address plugin);
    event PluginUninstalled(address plugin);

    function installPlugin(PluginInstallConfig calldata config) external;

>     function uninstallPlugin(PluginUninstallConfig calldata config) external;

Author: I chose to leave out `uninstallPluginWithManifest` here. I think this can be an optional function provided by the account, and I allude to this in the "Calls to `uninstallPlugin`" section below.

```solidity
interface IPluginLoupe {

>     struct ExecutionFunctionConfig {

Author: Renamed from ExecutionPluginConfig to match the function name.

        address plugin;
        FunctionReference userOpValidator;

>         FunctionReference runtimeValidator;

Author: Was there a reason these were implemented these as `FunctionReference[]` in our current code?

Reviewer 2: I think if we're moving away from "bare" functions to internal ones within `preExecHook`, etc then we should completely get rid of the `FunctionReference` type in the spec and update any corresponding return values in the loupe functions to be the underlying new type (address and one of `uintN` or `bytesN`).

It was a bit more ok to use `bytes24` = `FunctionReference` with the prior scheme, as the "correct" way to encode  `function` ABI types was also `bytes24`, so they were all aliasing the same underlying type. Since the `bytes4` portion now no longer references a function selector but a new concept, I think we should avoid coincidentally making it conform to the same convention as function selectors.

Execution interface for calls made from plugins. Modular Smart Contract Accounts MUST implement this interface to support execution from plugins.

```solidity
interface IPluginExecutor {
    function executeFromPlugin(bytes calldata data) external payable;

    function executeFromPluginExternal(address target, uint256 value, bytes calldata data) external payable;
}
```

The `executeFromPluginExternal` function SHOULD check whether the call's target implements the `IPlugin` interface via ERC-165. If the target is a plugin, the call SHOULD revert. This prevents accidental misconfiguration or misuse of plugins (both installed and uninstalled). Installed plugins MAY interact with other installed plugins via the `executeFromPlugin` function.

#### `IPlugin.sol`

Plugin interface. Plugins MUST implement this interface to support plugin management and interactions with MSCAs.

```solidity
interface IPlugin {
    /**
     * @notice Initialize plugin data for the modular account.
     * @dev Called by the modular account during `installPlugin`.
     * @param data Optional bytes array to be decoded and used by the plugin to setup initial plugin data for the modular account.

>      */

Reviewer 2: So far the convention within the standard has been to describe these in the text components, rather than the code block. I think we should prefer that for 2 reasons:

1. The spec should be language-agnostic, but natspec is a Solidity-exclusive concept. We've been embedding the concept of ABI-encoding into the spec rather heavily, but function signatures and ABI-encoding has generally widespread adoption across languages. We should keep these as interfaces to be more (but still not completely) language agnostic.
2. We can dedupe writing for some sections by describing how variables named the same represent the same concept - such as for pre user op validation hook params and user op validation function params, much like in the now-removed "function signatures" section.

Author: Yeah, makes sense - will update

Reviewer 2: Discussed offline - I'll stack a PR on top of this w/ this change

// For functions of type `ManifestFunctionType.DEPENDENCY`, the MSCA MUST find the plugin address
// of the function at `config.dependencies[dependencyIndex]` during the call to `installPlugin(config)`.
struct ManifestFunction {
    ManifestFunctionType functionType;
    uint32 functionId;
    uint256 dependencyIndex;
}

struct ManifestValidator {
    bytes4 executionSelector;
    ManifestFunction validator;
}

struct ManifestPreValidationHook {
    bytes4 executionSelector;
    ManifestFunction validator;
    ManifestFunction hook;
}

struct ManifestExecutionHook {
    bytes4 executionSelector;
    ManifestFunction preExecHook;
    ManifestFunction postExecHook;
}

struct ManifestStandardExecutionHook {
    ManifestFunction validator;
    ManifestFunction preExecHook;
    ManifestFunction postExecHook;
}

struct ManifestExternalCallPermission {
    address[] contracts;
    bytes4[] functionSelectors;
    bool permitAnyContract;
    bool permitAnySelector;
}

struct PluginManifest {
    // A human-readable name of the plugin.
    string name;
    // The version of the plugin, following the semantic versioning scheme.
    string version;

>     // The author field SHOULD be the GitHub handle of the user or organization that created this plugin.

Reviewer 2: Probably want to exclude specific platform identifiers from the spec. Also, since this is not a "secure" validation of authorship, we can be more lax on the spec's wording.

Author: Was following Yoav's suggestion here, but yeah agree the convention can probably be communicated via other venues.

```suggestion
    // The author field MAY be a username representing the identity of the user or organization that created this plugin.
```
?

Reviewer 2: looks good, can we keep `SHOULD`? It's logically the same content as `MAY` but communicates intent a bit more strongly

    string author;

> 

Author: I chose to leave out `string[] requiredFeatures;` for a future update.

    // If this plugin depends on other plugins' validation functions and/or hooks, the interface IDs of
    // those plugins MUST be provided here, with its position in the array matching the `dependencyIndex`
    // members of `ManifestFunction` structs used in the manifest.
    bytes4[] dependencyInterfaceIds;

    // Execution functions defined in this plugin to be installed on the MSCA.
    bytes4[] executionSelectors;
    // Native functions or execution functions already installed on the MSCA that this plugin will be
    // able to call.
    bytes4[] permittedExecutionSelectors;
    // External contract calls that this plugin will be able to make.
    ManifestExternalCallPermission[] permittedExternalCalls;

    ManifestValidator[] userOpValidators;
    ManifestValidator[] runtimeValidators;
    ManifestPreValidationHook[] preUserOpValidationHooks;
    ManifestPreValidationHook[] preRuntimeValidationHooks;
    ManifestExecutionHook[] executionHooks;
    ManifestStandardExecutionHook[] standardExecutionHooks;
}
```

### Expected behavior

#### Calls to `installPlugin`

> The function `installPlugin` accepts a parameter `PluginInstallConfig calldata config`, which contains the address of the plugin to install, the Keccak-256 hash of the plugin's manifest, RLP-encoded data to pass to the plugin's `onInstall` callback, and an array of addresses that represent the plugin's install dependencies.

Reviewer: are we using RLP encoding? i think solidity's abi.encode might be a diff type of encoding which isnt RLP

Author: Oops this should be ABI-encoded

- Revert if `config.manifestHash` does not match the computed Keccak-256 hash of the plugin's manifest. This prevents installation of plugins that attempt to install a different plugin configuration than the one that was approved by the client.
- Revert if any address in `config.dependencies` does not support the interface at its matching index in the manifest's `dependencyInterfaceIds`, or if the two array lengths do not match, or if any of the dependencies are not already installed on the modular account.

The function SHOULD store a record of the manifest hash and dependencies that were used for the plugin's installation. Each dependency's record SHOULD also be updated to reflect that it has a new dependent. These records facilitate the uninstall process.

The function MUST store the plugin's permitted execution selectors and external contract calls to be able to validate calls to `executeFromPlugin` and `executeFromPluginExternal`.

The function MUST parse through the execution functions, validation functions, and hooks in the manifest and add them to the modular account after resolving each `ManifestFunction` type.

> - Each execution selector MUST be added as a valid execution function on the modular account. If the execution selector has already been added or matches the selector of a native function, the function MUST revert.

Reviewer: i think we can weaken the condition here from MUST to SHOULD revert for the standard. for instance, honest selector collisions are probably really likely to happen in practice and some 6900 accounts might implement:
1. A `install --force` type of installation route to handle these selector collisions
2. Have additional "selector collision hooks" for fallback calls
3. during onInstall, accept user-specified selector collision hooks that perform the call routing logic to different selectors (e.g. routes are based on calldata length)

Author: True, will update!

Reviewer 2: This seems like a sufficiently edge-case usage scenario that I'd argue the standard should not make exceptions to support, and that it should actually revert. Given that execution selectors are inherently what cross-plugin logic depends on (cross-plugin validators and hooks), allowing any sort of mis-config here invalidates the integrity of plugin manifests.

Reviewer: another case is accounts allowing non 1:1 mapping from incoming msg.sig to outgoing msg.sig by default (i.e. calling `account.swapToUniswap()` can maps to `UniswapV2Router.swap()`). i think `SHOULD` makes a note that account devs need to think about collisions, and suggests 1 way of handling collisions which is to revert

that said, have thought about these other impls less than you so will defer to both of u on final decision

> - If an associated function that is to be added already exists, execution SHOULD NOT revert but continue to the next operation.

Reviewer 2: Ah I remember discussing this, this leads to installation-order-dependent account configs. This might change with our work on DI, so I think it's ok to leave in for now, but I'd argue this also falls under the "invalidates the integrity of plugin manifests" bucket.

Author: Ah yup. Unless we keep a reference count for each conflict or something to be able to cleanly uninstall... 😬

Finally, the function MUST call the plugin's `onInstall` callback with the data provided in `config.data`. This serves to initializes the plugin state for the modular account. If `onInstall` reverts, the `installPlugin` function MUST revert.

> Plugins MAY configure associated functions on the `installPlugin` function. If present, the validation function and hooks MUST be called in the appropriate sequence.

Reviewer 2: Given that the sequence has not yet been introduced in the document, maybe it makes sense to shift this entire section to after the sections titled "Calls to validateUserOp", "Calls to execution functions", and "Calls made from plugins"?

#### Calls to `uninstallPlugin`

The function `uninstallPlugin` accepts a parameter `PluginUninstallConfig calldata config`, which contains the address of the plugin to uninstall and ABI-encoded data to pass to the plugin's `onUninstall` callback.

The function MUST perform the following preliminary checks:

- Revert if the plugin is not installed on the modular account.

> - Revert if the hash of the manifest used at install time does not match the computed Keccak-256 hash of the plugin's current manifest. This prevents unclean removal of plugins that attempt to force a removal of a different plugin configuration than the one that was originally approved by the client for installation. To allow for removal of such plugins, the modular account MAY implement another uninstall function that accepts the original plugin manifest as a parameter.

Reviewer 2: This introduces a hard requirement on using account storage, despite the usage of the wording `SHOULD` in the previous section. For internal consistency, we should reword one of the two locations.

Personally, I'd prefer to remove this overall requirement, but I'm still working through what options we have and where standardization would help and/or make sense.

Author: Separated out this section under `SHOULD` for now.

- Revert if there is at least 1 other installed plugin that depends on execution functions, validation functions, or hooks added by this plugin.

> The function SHOULD remove any metadata stored for the plugin. Each dependency's record SHOULD also be updated to reflect that it has no longer has this plugin as a dependent.

Reviewer 2: Does this comment aim to address non-standard metadata?

Author: I was getting at anything stored to support IPluginLoupe, but perhaps this wording is too broad. Are you thinking some account impls may want to keep metadata around about which plugins have been historically installed?

Reviewer 2: I meant this more around "leaking" implementation details into the spec, but it looks like we use the term "metadata" to refer to storing the plugin manifest hash, so it's ok for the time being.

- Run the execution function.
- If any associated post execution hooks are defined, run the functions. If a pre execution hook returned data to the account, that data MUST be passed as a parameter to the associated post execution hook. The operation MUST be done with the `call` opcode. If any of these functions revert, execution MUST revert.

> **⚠️ If the execution function does not have a definition for either pre runtime validation hooks, pre execution hooks or post execution hooks, the undefined functions will be skipped. The execution function will be run and it may change account state.**

#### Calls made from plugins

Plugins MAY interact with other plugins and external contracts through the modular account using the special functions defined in the `IPluginExecutor` interface. These functions MAY be called without a defined validation function, but the modular account MUST enforce these checks and behaviors:

> The `executeFromPlugin` function MUST allow plugins to call functions (either native or execution functions installed by plugins) on the MSCA. Hooks matching the function selector provided in `data` and hooks associated with the function selector and the calling plugin address MUST be called. If the calling plugin's manifest did not include the provided function selector within `permittedExecutionSelectors` at the time of installation, execution MUST revert.

Reviewer: the way its worded here is interesting - does this mean we no longer store `_execFromPluginPermitted` and just check from the calling plugins manifest?

if that's the case we'll need to make sure that:
1. the calling plugin is installed, else revert
2. the calling plugins manifest still matches the storedManifest, else revert

Author: No I just wanted to abstract away how the account keeps track of this. But perhaps this is confusing.

Reviewer: maybe theres some room for improvement but knowing our current implementation probably biased me towards expecting a diff wording

The `executeFromPluginExternal` function MUST allow plugins to call external contracts as specified by its parameters on behalf of the MSCA. If the calling plugin's manifest did not explicitly allow the external contract call within `permittedExternalCalls` at the time of installation, execution MUST revert.

## Rationale

ERC-4337 compatible accounts must implement the `IAccount` interface, which consists of only one method that bundles validation with execution: `validateUserOp`. A primary design rationale for this proposal is to extend the possible functions for a smart contract account beyond this single method by unbundling these and other functions, while retaining the benefits of account abstraction.

The function routing pattern of ERC-2535 is the logical starting point for achieving this extension into multi-functional accounts. It also meets our other primary design rationale of generalizing execution calls across multiple implementing contracts. However, a strict diamond pattern is constrained by its inability to customize validation schemes for specific execution functions in the context of `validateUserOp`, and its requirement of `delegatecall`.

> This proposal includes several interfaces that build on ERC-4337 and are inspired by ERC-2535. First, we standardize a set of modular functions that allow smart contract developers greater flexibility in bundling validation, execution, and hook logic. We also propose the `IPluginLoupe` interface that takes inspiration from the diamond standard and provides methods for querying execution functions, validation functions, and hooks.

Author: Gotcha, I thought the purpose of this paragraph was to specifically call out interfaces inspired by ERC-2535 so only left the relevant one in. I can expand to cover others.

> This proposal includes several interfaces that build on ERC-4337 and are inspired by ERC-2535. First, we standardize a set of modular functions that allow smart contract developers greater flexibility in bundling validation, execution, and hook logic. We also propose the `IPluginLoupe` interface that takes inspiration from the diamond standard and provides methods for querying execution functions, validation functions, and hooks.

Reviewer: small nit: inconsistency here in the conclusion vs header which states "this proposal standardizes smart contract accounts and account plugins, which are smart contract interfaces that allow for composable logic within smart contract accounts"
